### PR TITLE
fix(codex): bind MCP helpers to managed account homes

### PR DESCRIPTION
**Pull request by seonghobae**
## ELI5

Managed Codex accounts can share an MCP header helper while still passing the correct account home to that helper. This lets the helper select the matching account identity instead of losing the account context when Orca copies the shared configuration into a managed home.

## What Changed

Before, Orca copied `mcp_servers.*.http_headers_helper` into each managed Codex home unchanged, so a shared helper had no reliable account-home argument.

After, a shared helper may include a standalone `{{ORCA_MANAGED_CODEX_HOME}}` argument. Orca replaces it with the validated managed home while mirroring configuration and while preparing host and WSL launches. The system template and unrelated settings remain unchanged.

The binder quotes literal POSIX paths and leaves ambiguous templates unresolved. Quoted or embedded placeholders, dollar and command expansions, backticks, shell control syntax, line continuations, heredocs, and native Windows paths fail closed. WSL UNC homes are converted to their Linux paths before binding.

## Why

Native MCP helpers can execute outside the managed Codex process environment. Passing the validated account home as an explicit argument preserves account isolation without silently substituting a default account. Restricting replacement to a standalone argument avoids turning account-home text into shell syntax.

## Linked Issue

Related: #20741 describes managed Codex homes losing MCP capabilities and records explicit MCP registration as a workaround. This PR addresses account-aware helper binding and does not close that broader bundled Chrome/CUA issue. No separate issue was opened before this PR.

## Visual Proof

N/A — this changes managed Codex configuration mirroring and MCP helper arguments; it has no visual or interaction change.

## Testing

- [x] Automated tests added/updated, including binder validation, real-file mirroring, host/WSL launch wiring, platform behavior, and a gated native Codex app-server integration fixture
- [x] `pnpm typecheck:cli`
- [x] 55 targeted tests across the binder, real-file mirror, and `prepareForCodexLaunch` lifecycle at `d4055c627183cb196a1bc3e7c08adaa3b4878565`
- [x] POSIX literal-argument verification with one accepted template and 12 rejected ambiguous templates
- [x] Two native Codex app-server clients exercised account-separated MCP calls against an identical-source audited gateway
- [ ] I manually tested these changes locally through an installed Orca build

Platforms covered: macOS/POSIX behavior was exercised directly. Automated tests cover WSL UNC conversion and launch wiring. Native Windows intentionally leaves the placeholder unresolved and has platform-specific regression coverage.

## AI Disclosure

OpenAI Codex and Claude assisted implementation, tests, and review. Human review reproduced the shell-boundary cases and verified the native-client account separation.

## Review

The code review checked:

- Cross-platform behavior: POSIX quoting, WSL UNC-to-Linux conversion, and native Windows fail-closed behavior are explicit and tested.
- Local and SSH/remote behavior: the change is limited to managed Codex home mirroring and launch preparation; it does not assume the workspace is local.
- Agent and integration compatibility: behavior is guarded to Codex managed homes and `mcp_servers.*.http_headers_helper`; unrelated agents, MCP entries, and settings are preserved.
- Performance: binding parses and rewrites configuration only during existing mirror/launch preparation paths; it adds no hot-path polling or background work.
- UI quality: not applicable because there is no renderer or interaction change.
- Security: only a validated managed home is substituted, unsafe shell contexts remain unresolved, and there is no default-account fallback.

Latest automated review at `7691ea7e817037d62bfc17684b34e298960b73fa` reported no new issues. Upstream CI and maintainer review remain pending.

## Agent skill upstream boundary

- [x] Not applicable; this change copies or mechanically translates no upstream skill-installer source, tests, fixtures, registry entries, path tables, comments, or documentation.

## Author

X: [@Personnelpsyc](https://x.com/Personnelpsyc)

## Notes

This PR is a source candidate. It does not patch or deploy an installed Orca application. PR #21665 is unrelated and unchanged.

## Checklist

- [x] This PR is small and focused
- [x] I explained what changed and why (ELI5, the user-facing before/after, the mechanism, and why over the alternatives)
- [x] Before/after screenshots or videos attached for UI changes, or `N/A` with reason
- [x] Self-reviewed for correctness, security, and performance
- [x] Cross-platform, SSH/remote, and path/shortcut impact considered (or N/A)
- [ ] `pnpm lint`, `pnpm typecheck`, `pnpm test`, and `pnpm build` pass (targeted tests and CLI typecheck pass; full local suite/build not claimed)

### CODE REVIEW

**Review by seonghobae — COMMENTED** [review]
Lead review of 11f5f548e65a2d3ece13e61a71beb50e2656e716 reproduced unsafe quote-context substitution with the exact source binder and a real POSIX shell. A helper containing `printf '[%s]' "{{ORCA_MANAGED_CODEX_HOME}}"` and a home containing literal `$(printf injected)` executes that substitution instead of passing the original path. The bare placeholder case passes correctly.

Please preserve the complete literal argument or reject unsupported quoted/embedded contexts without substituting. A real argv round-trip test should cover apostrophes, dollar substitution and backticks; checking only the encoded TOML substring does not catch this. Follow-up repair is already assigned. This head is not accepted for deployment.

**Review by seonghobae — COMMENTED** [review]
Rechecked exact head dd7ef67bc1e4ee3428438467061af120d38615da against both prior quote-context reproductions. A real /bin/sh literal-argv test preserves spaces, apostrophe, dollar substitution text, backticks and variable-like text in the home. Ten unsupported templates remain unexpanded, including quoted/embedded markers, command substitution, arithmetic/braced expansion, line continuation and heredoc. Native Windows remains unsupported; no Windows consumer acceptance is claimed.

At this exact binder revision, two fresh installed Codex app-server processes using two real managed account homes called get_current_time and convert_time successfully (four actual returns; 18:12:43 and18:12:48 KST). An audit-only wrapper around the unchanged gateway server recorded distinct authenticated principals on their actual native tools/call sessions. Each live session returned200 to its owner and404 to the other account. Observed gateway/client PIDs78091/79077/82133 had parent77782; both clients had no time-server descendants before or after calls. All owned temporary processes were cleaned and the harness exited0. No model inference was used.

The reported quote-expansion findings are resolved in this tested scope. This is controlled source/native-client evidence, not installed Orca release integration, qualifying GitHub approval, or an LLM-driven end-to-end result. The existing account usage-limit failure remains separate. No merge or installed-app deployment was performed.

File: src/main/codex/codex-managed-home-mcp-binding.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  platform: NodeJS.Platform = process.platform
+): string {
+  const lines = config.split('\n')
+  let table = ''
+  let scanState = createTomlLineScanState()
+  const wslHome = parseWslUncPath(managedHomePath)?.linuxPath
+  const helperHome = wslHome
+    ? quotePosixShell(wslHome)
+    : platform === 'win32'
+      ? quoteWindowsCmdArgument(managedHomePath)
```
[the remark this answers is not in the document]
Comment (seonghobae) [reply src/main/codex/codex-managed-home-mcp-binding.ts]: Addressed in the current source: native Windows returns the unexpanded configuration before any home quoting, so this feature does not claim cmd.exe or delayed-expansion support. The native-Windows refusal regression stays enabled; no Windows argument-consumer acceptance is claimed.

File: src/main/codex/codex-managed-home-mcp-binding.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      continue
+    }
+
+    const char = command[index] ?? ''
+    const next = command[index + 1]
+    if (
+      char === '\n' ||
+      char === '\r' ||
+      char === '`' ||
+      char === '$' ||
```
[the remark this answers is not in the document]
Comment (seonghobae) [reply src/main/codex/codex-managed-home-mcp-binding.ts:89]: This is an intentional conservative opt-in contract for now: template commands containing dollar characters are unsupported, even inside single quotes; safely quoted literal dollar characters introduced through the managed-home argument remain supported and have a real /bin/sh regression. Existing helpers without the new placeholder are unchanged. Supporting additional shell-template forms can be a separate change with consumer tests; this patch does not broaden the grammar.

**Review by seonghobae — COMMENTED** [review]
Reviewed exact head d4055c627183cb196a1bc3e7c08adaa3b4878565. The escaped-space/tab guard is correct; the revised regression now uses plain printf without unrelated semicolon/dollar rejection. The CLI project inclusion repairs TS6307. MacBookAir owner reports 55 tests across binder, real-file mirror and prepareForCodexLaunch lifecycle, plus passing CLI typecheck.

Independent lead verification at this exact head: real /bin/sh literal-home positive and 12 negative templates passed. Two installed native Codex app-server clients each called get_current_time and convert_time against the unchanged gateway source with an audit-only ASGI wrapper. Server authentication observed distinct account principals and native tools/call sessions. Each live own session returned200; the other account session returned404 in both directions. PID/parent identity recorded; no time backend descendants before/after. Model turns0; proof process exit0 and owned temporary processes cleaned.

No further defect found within this reviewed scope. Native Windows binding is intentionally unsupported. These controlled source/transport results do not establish installed Orca new-session configuration regeneration or deployment acceptance. This COMMENT review is not a qualifying GitHub approval; normal upstream checks/review remain required.

**Comment by seonghobae** [comment]
Launch follow-up is now at `69fd9642fc82450db1966d818d339bff1af1a63c`.

The final launch-time config regeneration was still dropping the explicit account binding on both host and WSL paths. Both callers now pass the selected managed home to the existing mirror binder. The WSL regression fails with the old caller and also checks that the system-default lane does not mirror.

Validation: 21 tests passed, zero skipped, including a real native Codex app-server session after launch preparation; Node typecheck, changed-file lint and formatting passed. The native probe now uses the shared process runner and fails on an installed executable error instead of silently skipping it.

Scope: the native test forwards a disposable principal through initialize/list traffic; outer Electron and hook services are mocked. This does not establish packaged Orca installation or authenticated Gateway tools/call acceptance. Those require the isolated artifact and two-principal session checks. Independent review should target this new head.

**Comment by seonghobae** [comment]
Fixture review follow-up is pushed at `5803333913b2582d6e4775057f28a3825913b906`. The native helper test is explicitly POSIX-only; listener setup is covered by cleanup, and cleanup failures fail the test while preserving simultaneous primary failures. Direct verbose native test passed (1/1); worker reports direct Node typecheck and lint success. Prior 21-test evidence belongs to `69fd9642`, not a repeated full run on this head. Packaged Orca and authenticated Gateway two-principal tools/call acceptance remain unverified. Please review this final head.

**Comment by seonghobae** [comment]
@coderabbitai review

Please review final head 5803333913b2582d6e4775057f28a3825913b906. Automatic reviews are paused and the displayed prior review targets an older head. Launch mirror and fixture follow-ups are documented above; packaged application and live Gateway acceptance remain unverified.

**Comment by seonghobae** [comment]
Fixed the outstanding platform test issue at `7691ea7e817037d62bfc17684b34e298960b73fa`.

The original config-mirror test now asserts native Windows leaves the helper placeholder unresolved, matching the intentional fail-closed contract. Supported POSIX paths still require separate account bindings. Added a Windows-negative mirror regression and a WSL UNC-to-POSIX binder case.

Validation: the original case failed under forced `win32` before the correction and passed afterward; the final two-file suite passed all 46 tests. Changed-file lint passed per executor receipt. Lead reviewed the diff and RED/GREEN logs. This is simulated platform-contract coverage, not a native Windows runtime acceptance claim.

Please re-review https://github.com/stablyai/orca/pull/21771#discussion_r4056620794 against this head. Packaged Orca launch/regeneration remains unverified. A separate fresh live Gateway SDK check passed initialize/list/call for two principals and rejected both foreign-session attempts with 404; it does not establish installed Orca integration.

**Comment by brennanb2025** [comment]
Thanks for putting this together. We traced the underlying problem to Codex itself: it starts `http_headers_helper` without `CODEX_HOME`, so the helper can't tell which Codex home it's serving (details in #22471). Rather than having Orca rewrite commands in the mirrored config, we're waiting for Codex to adopt the proper behavior, which we've requested in https://github.com/openai/codex/issues/46244#issuecomment-5822540558. We'll leave this open for now and revisit once that lands.
